Add invalid attempts

iamazzeez opened the pull request:
https://chefio.atlassian.net/browse/KINETICS-66
https://chefio.atlassian.net/browse/KINETICS-67
https://chefio.atlassian.net/browse/KINETICS-68
https://chefio.atlassian.net/browse/KINETICS-79
https://chefio.atlassian.net/browse/KINETICS-73

<!--
Thank you for sending a pull request! Here are some tips for contributors:

1. Fill the description template below.
2. Sign a DCO (if you haven't already signed it).
3. Include appropriate tests (if necessary). Make sure that all CI checks passed.
4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.
-->

#### Overview

<!-- Describe your changes briefly here. -->

#### What this PR does / why we need it
This PR creates blocked_user table in dex psql db
This is needed to accommodate failed login attempts with valid username 

<!--
- Please state in detail why we need this PR and what it solves.
- If your PR closes some of the existing issues, please add links to them here.
  Mentioned issues will be automatically closed.
  Usage: "Closes #<issue number>", or "Closes (paste link of issue)"
-->

#### Special notes for your reviewer

#### Does this PR introduce a user-facing change?

<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->

```release-note

```

iamazzeez committed:
added logs

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
added blocked_user talble

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
createblockeduser type fix

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
added test cases

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
added test cases

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
added kubernetes storage code

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
added etcd storage code

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
migrate stmts

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
migrate only for psql

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
migrate only for psql

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
migration fix

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
test

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
test

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
check if user is blocked logic

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
migration syntax fix

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
wip blocked user logic

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
update counter syntax fix

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
added ui render logic

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
added ui render logic fix

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
check if user is blocked condition

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
check if user is blocked condition fix

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
check if user is blocked condition fix

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
check if user is blocked condition fix

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
delete user on successful login

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
delete user on successful login fix

Signed-off-by: Abdul-Az <aazeez@progress.com>

kalroy commented on storage/kubernetes/storage.go:
```diff
@@ -307,6 +321,20 @@ func (cli *client) getPassword(email string) (Password, error) {
 }
 
+func (cli *client) getBlockedUser(username string) (BlockedUser, error) {
+	// TODO(ericchiang): Figure out whose job it is to lowercase emails.
```
this comment??

Abdul-Az replied:
removed it.

iamazzeez committed:
add config from automate

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
typo fix

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
added invalidAttempts config to server

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
added invalidAttempts config to server fix

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
maxAllowedAttempts type fix

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
maxAllowedAttempts send to ui

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
maxAllowedAttempts send to ui fix

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
blockDuration send to ui fix

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
removed unnecessary commits

Signed-off-by: Abdul-Az <aazeez@progress.com>

Abdul-Az commented on server/handlers.go:
```diff
@@ -392,25 +392,110 @@ func (s *Server) handleConnectorLogin(w http.ResponseWriter, r *http.Request) {
+				isBlockedUserUpdated := false
+				// Since user has exceeded the specified blocked duration reset the
+				// counter and updated_at time
+				if diff := time.Since(blockedUser.UpdatedAt); diff.Minutes() > 30 {
```
got it, will update it.

iamazzeez committed:
code refactored
'

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
code refactor

Signed-off-by: Abdul-Az <aazeez@progress.com>

Abdul-Az commented on server/handlers.go:
```diff
@@ -392,25 +392,110 @@ func (s *Server) handleConnectorLogin(w http.ResponseWriter, r *http.Request) {
+				UpdatedAt:            time.Now(),
+			})
+			if err == storage.ErrAlreadyExists {
+				blockedUser, err := s.storage.GetBlockedUser(username)
```
code refactored

Dmaddu commented on storage/sql/crud.go:
```diff
@@ -516,6 +516,32 @@ func (c *conn) UpdateClient(id string, updater func(old storage.Client) (storage
 	})
 }
 
+func (c *conn) UpdateBlockedUser(username string, updater func(old storage.BlockedUser) (storage.BlockedUser, error)) error {
```
instead of UpdateBlockedUser & CreateBlockedUser can't we write one method for upsert?

Abdul-Az replied:
We can, but in terms of code readability update and create looks correct.

Dmaddu commented on storage/sql/migrate.go:
```diff
@@ -274,4 +274,15 @@ var migrations = []migration{
+	{
+		stmts: []string{
+			`
+			create table if not exists blocked_user (
```
how about invalid_login_attempts as table name instead of blocked_user?

Abdul-Az replied:
seems good, will update it.

Dmaddu commented on server/handlers.go:
```diff
@@ -209,6 +209,53 @@ func (s *Server) discoveryHandler() (http.HandlerFunc, error) {
 	}), nil
 }
 
+func (s *Server) isFailedAttemptDurationExceeded(u storage.BlockedUser) bool {
```
definitely, this code is not easy to maintain, at least the terminology is not proper.
FYI - we are using the BlockedUser instance at most of the places, but that is not a blocked user, that is just an invalid attempt, user will be blocked when he exceeds the max invalid attempts. 

Either please change the terminology or provide more comments.

Abdul-Az replied:
Sure, will change the terminology

Dmaddu reviewed (changes requested):
please re-look into the code in terms of a maintenance perspective. Add more comments and use proper terminology to understand the code better.

Abdul-Az commented on storage/static.go:
```diff
@@ -161,6 +161,68 @@ func (s staticPasswordsStorage) UpdatePassword(email string, updater func(old Pa
+		// Enable case insensitive username comparison.
+		lowerUsername := strings.ToLower(u.Username)
+		if _, ok := blockedUsersByUsername[lowerUsername]; ok {
+			logger.Errorf("Attempting to create StaticblockedUsers with the same username: %s", u.Username)
```
it is just for printing the error on logs.

iamazzeez committed:
renamed BlockedUser to InvalidLoginAttempt

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
config terminology update.

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
sql row name update

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
debug stmts

Signed-off-by: Abdul-Az <aazeez@progress.com>

Dmaddu commented on server/handlers.go:
```diff
@@ -392,25 +439,88 @@ func (s *Server) handleConnectorLogin(w http.ResponseWriter, r *http.Request) {
+				return
+			}
+
+			s.updateInvalidAttemptCount(username, w, r)
```
what if the method returns any error?

Dmaddu commented on server/handlers.go:
```diff
@@ -209,6 +209,53 @@ func (s *Server) discoveryHandler() (http.HandlerFunc, error) {
 	}), nil
 }
 
+func (s *Server) isFailedAttemptDurationExceeded(u storage.InvalidLoginAttempt) bool {
```
how about "isBlockedTimeExpired"?

Dmaddu commented on server/handlers.go:
```diff
@@ -209,6 +209,53 @@ func (s *Server) discoveryHandler() (http.HandlerFunc, error) {
+	return false
+}
+
+func (s *Server) resetFailedAttempt(username string, w http.ResponseWriter, r *http.Request) {
```
I don't think there is a use case for this, isn't it?

iamazzeez committed:
removed logs

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
enable/disable invalid_login_attempts

Signed-off-by: Abdul-Az <aazeez@progress.com>

kalroy commented on server/handlers.go:
```diff
@@ -392,25 +439,88 @@ func (s *Server) handleConnectorLogin(w http.ResponseWriter, r *http.Request) {
+			}
+
+			if s.isFailedAttemptDurationExceeded(InvalidLoginAttempt) {
+				s.resetFailedAttempt(username, w, r)
```
The question remains on how are we handling the scenario where a user may have failed login for less than the number of attempts in the duration and make another failed login. Eg: In 30 mins duration the user made 4 invalid attempts. After 30 mins if he tries again and it is an invalid attempt, the user should not be locked out. His attempt count should start from 1 again.

kalroy reviewed (commented):
Check the scenario answered for the reset attempt

iamazzeez committed:
added logs

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
removeed logs

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
ldap invalidattempts handle fix

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
handle duplicate entries in invalid_attempts table

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
handleInvalidLoginAttempts update for ldap

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
error fix

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
error fix

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
changed blockedDuration to blockedDurationInMinutes

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
reverted if not exists from create table

Signed-off-by: Abdul-Az <aazeez@progress.com>

Dmaddu committed:
Merge pull request #53 from chef/ldap_fix

Fixed errors occurred while testing

iamazzeez committed:
updated_at time

Signed-off-by: Abdul-Az <aazeez@progress.com>

iamazzeez committed:
updatedat fix

Signed-off-by: Abdul-Az <aazeez@progress.com>

Dmaddu committed:
Merge pull request #56 from chef/ldap_fix

Ldap fix

SanjuPal01 committed:
Sanju/test refresh (#81)

* Testing Refresh API

Signed-off-by: SanjuPal01 <sanju.sanju@progress.com>

* Change 2

* Handler

* Last Used

* test

* Time Difference

* Check

* Token Validator

Signed-off-by: SanjuPal01 <sanju.sanju@progress.com>

* Checking

* Check

* Test

* Test

* Check

* Check

* Check

* Check

* Check

* Check

* Check

* Check

* Check

* Check

* Check

* Removing Redundant Code

Signed-off-by: SanjuPal01 <sanju.sanju@progress.com>

* Token Validator

Signed-off-by: SanjuPal01 <sanju.sanju@progress.com>

* Removing Redundant Code

Signed-off-by: SanjuPal01 <sanju.sanju@progress.com>

* Code Updated

Signed-off-by: SanjuPal01 <sanju.sanju@progress.com>

* Fixed validator

Signed-off-by: SanjuPal01 <sanju.sanju@progress.com>

* naming convention

Signed-off-by: SanjuPal01 <sanju.sanju@progress.com>

iamazzeez committed:
pq pkg upgrade (#151)

Signed-off-by: Abdul-Az <aazeez@progress.com>

YashviJain01 committed:
Update packages (#270)

* Update go version

Signed-off-by: Yashvi Jain <Yashvi.jain@progress.com>

* Adding the go version as 1.24.3

Signed-off-by: Yashvi Jain <Yashvi.jain@progress.com>

* Adding new version of goxmldsig

Signed-off-by: Yashvi Jain <Yashvi.jain@progress.com>

* Changing thr sqaure/go-hose package to go-jose/go-jose

Signed-off-by: Yashvi Jain <Yashvi.jain@progress.com>

* Upgrading go

Signed-off-by: Yashvi Jain <Yashvi.jain@progress.com>

* Fixing the pipeline

Signed-off-by: Yashvi Jain <Yashvi.jain@progress.com>

* Adding the test case and pipeline fix

Signed-off-by: Yashvi Jain <Yashvi.jain@progress.com>

* Upgrading go conatienrs package

Signed-off-by: Yashvi Jain <Yashvi.jain@progress.com>

* upgrade test-conatiner package evrsion

Signed-off-by: Yashvi Jain <Yashvi.jain@progress.com>

* Adding updated packages

Signed-off-by: Yashvi Jain <Yashvi.jain@progress.com>

* Updating test-containers

Signed-off-by: Yashvi Jain <Yashvi.jain@progress.com>

* Updating packages

Signed-off-by: Yashvi Jain <Yashvi.jain@progress.com>

* Commenting ldap tests

Signed-off-by: Yashvi Jain <Yashvi.jain@progress.com>

* commenting kubernetes test cases

Signed-off-by: Yashvi Jain <Yashvi.jain@progress.com>

* Remvoing old packages to upgrade go version

Signed-off-by: Yashvi Jain <Yashvi.jain@progress.com>

* Solving lint issues

Signed-off-by: Yashvi Jain <Yashvi.jain@progress.com>

* Removing verify proto from the pipeline

Signed-off-by: Yashvi Jain <Yashvi.jain@progress.com>

* Removing duplicate entries

Signed-off-by: Yashvi Jain <Yashvi.jain@progress.com>

---------

Signed-off-by: Yashvi Jain <Yashvi.jain@progress.com>
Co-authored-by: Yashvi Jain <Yashvi.jain@progress.com>

YashviJain01 committed:
upgrade clientgolang package (#271)

* upgrade clientgolang package

Signed-off-by: Yashvi Jain <Yashvi.jain@progress.com>

* Adding nolint for wanted warnings

Signed-off-by: Yashvi Jain <Yashvi.jain@progress.com>

---------

Signed-off-by: Yashvi Jain <Yashvi.jain@progress.com>
Co-authored-by: Yashvi Jain <Yashvi.jain@progress.com>

Aishwarya2001A committed:
Update google.golang.org/grpc to v1.79.3

- Update grpc dependency from v1.29.0 to v1.79.3 in go.mod
- Remove storage/etcd backend (incompatible with grpc v1.30+ due to removed grpc/naming package)
- Remove etcd import and storage entry from cmd/dex/config.go
- Run go mod tidy to update go.sum

Signed-off-by: Aishwarya2001A <aaishwar@progress.com>

Aishwarya2001A committed:
Update Dockerfile to use Go 1.24 and Alpine 3.21

Signed-off-by: Aishwarya2001A <aaishwar@progress.com>

Aishwarya2001A committed:
Fix sqlite3 CGO build on Alpine with musl-libc by adding _LARGEFILE64_SOURCE

Signed-off-by: Aishwarya2001A <aaishwar@progress.com>

Aishwarya2001A committed:
Update CI: upgrade actions/setup-go to v5, remove etcd service

Signed-off-by: Aishwarya2001A <aaishwar@progress.com>

Aishwarya2001A committed:
Replace deprecated grpc.Dial with grpc.NewClient in api_test.go

Signed-off-by: Aishwarya2001A <aaishwar@progress.com>

Aishwarya2001A committed:
Restore etcd storage using go.etcd.io/etcd/client/v3 v3.5.21

Migrate from the old monolithic go.etcd.io/etcd module (which depended
on google.golang.org/grpc/naming, removed in grpc v1.30.0) to the
modern modular go.etcd.io/etcd/client/v3 v3.5.21.

Changes:
- storage/etcd/: restored all 4 files with updated import paths
  - go.etcd.io/etcd/clientv3 -> go.etcd.io/etcd/client/v3
  - go.etcd.io/etcd/clientv3/namespace -> go.etcd.io/etcd/client/v3/namespace
  - go.etcd.io/etcd/pkg/transport -> go.etcd.io/etcd/client/pkg/v3/transport
- cmd/dex/config.go: restored etcd storage registration
- go.mod: added go.etcd.io/etcd/client/v3 and client/pkg/v3 v3.5.21
- .github/workflows/ci.yml: update etcd image to v3.5.21, add DEX_ETCD_ENDPOINTS env var

Signed-off-by: Aishwarya2001A <aaishwar@progress.com>

Aishwarya2001A committed:
Fix etcd health check: use curl instead of etcdctl

etcdctl is not bundled in gcr.io/etcd-development/etcd:v3.5.21 image.
Switch health check to use curl against the /health HTTP endpoint.

Signed-off-by: Aishwarya2001A <aaishwar@progress.com>

Aishwarya2001A committed:
Fix etcd health check: revert to etcdctl (available in gcr.io/etcd-development/etcd image)

Signed-off-by: Aishwarya2001A <aaishwar@progress.com>

Aishwarya2001A committed:
Fix etcd CI: revert to gcr.io/etcd-development/etcd:v3.2.9

v3.2.9 was the original working image with a compatible etcdctl
health check. The etcd client/v3 Go module is still v3.5.21 —
the CI image version does not need to match the Go module version.

Signed-off-by: Aishwarya2001A <aaishwar@progress.com>

Aishwarya2001A committed:
fix(lint): fix gci import ordering in etcd storage files

Signed-off-by: Aishwarya2001A <aaishwar@progress.com>

kalroy committed:
Merge pull request #274 from chef/update-grpc-v1.79.3

Update google.golang.org/grpc to v1.79.3
